servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SetDynamicCompositeCancellable.java

```diff
@@ -81,14 +78,16 @@ public void cancel() {
 
     @Override
     public boolean add(Cancellable toAdd) {
-        final Set<Cancellable> currentSet = set;
-        if (currentSet == null) {
+        if (counter < 0) { // make best effort to avoid the set if cancelled already
             toAdd.cancel();
             return false;
-        } else if (!currentSet.add(toAdd)) {
+        } else if (!set.add(toAdd)) {
             return false; // user has implemented equals/hashCode so there is overlap?
-        } else if (!setUpdater.compareAndSet(this, currentSet, currentSet)) {
-            if (currentSet.remove(toAdd)) {
+        } else if (counterUpdater.incrementAndGet(this) <= 0) {
```

Comment:
consider plain `.get()`

Reply:
The IAG (or some atomic update) is necessary to ensure this thread observes a cancellation that raced with our above queue addition.